src/ringct/multiexp.cc

```diff
@@ -571,43 +577,33 @@ size_t get_pippenger_c(size_t N)
   return 9;
 }
 
-struct pippenger_cached_data
-{
-  size_t size;
-  ge_cached *cached;
-  pippenger_cached_data(): size(0), cached(NULL) {}
-  ~pippenger_cached_data() { aligned_free(cached); }
-};
-
 std::shared_ptr<pippenger_cached_data> pippenger_init_cache(const std::vector<MultiexpData> &data, size_t start_offset, size_t N)
 {
   MULTIEXP_PERF(PERF_TIMER_START_UNIT(pippenger_init_cache, 1000000));
   CHECK_AND_ASSERT_THROW_MES(start_offset <= data.size(), "Bad cache base data");
   if (N == 0)
     N = data.size() - start_offset;
   CHECK_AND_ASSERT_THROW_MES(N <= data.size() - start_offset, "Bad cache base data");
-  std::shared_ptr<pippenger_cached_data> cache(new pippenger_cached_data());
+  std::shared_ptr<pippenger_cached_data> cache = std::make_shared<pippenger_cached_data>();
 
-  cache->size = N;
-  cache->cached = (ge_cached*)aligned_realloc(cache->cached, N * sizeof(ge_cached), 4096);
-  CHECK_AND_ASSERT_THROW_MES(cache->cached, "Out of memory");
+  cache->resize(N);
   for (size_t i = 0; i < N; ++i)
-    ge_p3_to_cached(&cache->cached[i], &data[i+start_offset].point);
+    ge_p3_to_cached(&(*cache)[i], &data[i+start_offset].point);
 
   MULTIEXP_PERF(PERF_TIMER_STOP(pippenger_init_cache));
   return cache;
 }
 
 size_t pippenger_get_cache_size(const std::shared_ptr<pippenger_cached_data> &cache)
 {
-  return cache->size * sizeof(*cache->cached);
+  return cache->size() * sizeof(ge_cached);
 }
 
-rct::key pippenger(const std::vector<MultiexpData> &data, const std::shared_ptr<pippenger_cached_data> &cache, size_t cache_size, size_t c)
+ge_p3 pippenger_p3(const std::vector<MultiexpData> &data, const std::shared_ptr<pippenger_cached_data> &cache, size_t cache_size, size_t c)
```

Comment:
This definition doesn't exactly match what's in the header because of the lack of `const` qualifiers, so it might produce warnings on some compilers.

Reply:
Good catch, thanks